src/symbolic/symbolic_choice.ml

```diff
@@ -161,19 +158,21 @@ end = struct
       | Some f -> begin
         handle_status (Schedulable.run f wls) sched;
         Wq.end_pledge sched.work_queue;
-        work wls sched
+        work wls sched callback
       end
 
-    let spawn_worker sched wls_init =
-      Wq.make_pledge sched.res_writer;
+    let spawn_worker sched wls_init callback callback_init callback_close =
+      callback_init ();
       Domain.spawn (fun () ->
           Fun.protect
-            ~finally:(fun () ->
-              Wq.end_pledge sched.res_writer;
-              Wq.fail sched.work_queue )
+            ~finally:(fun () -> callback_close ())
             (fun () ->
               let wls = wls_init () in
-              work wls sched ) )
+              try work wls sched callback
+              with e ->
+                let bt = Printexc.get_raw_backtrace () in
+                Wq.fail sched.work_queue;
```

Comment:
You can put this in the `~finally` and remove the `try ... with` along with the exeception backtrace stuff